source/adapters/opencl/context.hpp

```diff
@@ -10,9 +10,74 @@
 #pragma once
 
 #include "common.hpp"
+#include "device.hpp"
 
-namespace cl_adapter {
-ur_result_t
-getDevicesFromContext(ur_context_handle_t hContext,
-                      std::unique_ptr<std::vector<cl_device_id>> &DevicesInCtx);
-}
+#include <vector>
+
+struct ur_context_handle_t_ {
+  using native_type = cl_context;
+  native_type Context;
+  std::vector<ur_device_handle_t> Devices;
+  uint32_t DeviceCount;
+  std::atomic<uint32_t> RefCount = 0;
+
+  ur_context_handle_t_(native_type Ctx, uint32_t DevCount,
+                       const ur_device_handle_t *phDevices)
+      : Context(Ctx), DeviceCount(DevCount) {
+    for (uint32_t i = 0; i < DeviceCount; i++) {
+      Devices.emplace_back(phDevices[i]);
+      urDeviceRetain(phDevices[i]);
+    }
+    RefCount = 1;
+  }
+
+  uint32_t incrementReferenceCount() noexcept { return ++RefCount; }
+
+  uint32_t decrementReferenceCount() noexcept { return --RefCount; }
+
+  uint32_t getReferenceCount() const noexcept { return RefCount; }
+
+  static ur_result_t makeWithNative(native_type Ctx, uint32_t DevCount,
+                                    const ur_device_handle_t *phDevices,
+                                    ur_context_handle_t &Context) {
+    try {
+      uint32_t CLDeviceCount;
+      CL_RETURN_ON_FAILURE(clGetContextInfo(Ctx, CL_CONTEXT_NUM_DEVICES,
+                                            sizeof(CLDeviceCount),
+                                            &CLDeviceCount, nullptr));
+      std::vector<cl_device_id> CLDevices(CLDeviceCount);
+      CL_RETURN_ON_FAILURE(clGetContextInfo(Ctx, CL_CONTEXT_DEVICES,
+                                            sizeof(CLDevices), CLDevices.data(),
+                                            nullptr));
+      if (DevCount != CLDeviceCount) {
+        return UR_RESULT_ERROR_INVALID_CONTEXT;
+      }
+      for (uint32_t i = 0; i < DevCount; i++) {
+        if (phDevices[i]->get() != CLDevices[i]) {
+          return UR_RESULT_ERROR_INVALID_CONTEXT;
+        }
+      }
+      auto URContext =
+          std::make_unique<ur_context_handle_t_>(Ctx, DevCount, phDevices);
+      Context = URContext.release();
+      CL_RETURN_ON_FAILURE(clRetainContext(Ctx));
```

Comment:
I think for native handles we should mirror l0's behaviour, that's where `isNativeHandleOwned` was first implemented so we can assume it reflects the original intention. The UR handles don't retain the native handle, and `isNativeHandleOwned` decides whether UR will release the native handle when the UR handle is released.

So for example l0 context tracks whether it owns its native handle (the param gets passed from [here](https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/context.cpp#L155) to [here](https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/context.hpp#L41)) and only uses that information when it comes to releasing the context [here](https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/context.cpp#L369)

Reply:
here's a draft for the new spec around native handles https://github.com/oneapi-src/unified-runtime/pull/2193